File: python/patterns-use-cases/src/dataupload/data_upload_service.py
import restate
import logging
from restate import WorkflowContext, WorkflowSharedContext, Workflow

from src.dataupload.utils import upload_data, create_s3_bucket, send_email

logging.basicConfig(level=logging.INFO, format='[%(asctime)s] [%(process)d] [%(levelname)s] - %(message)s')

data_upload_service = Workflow("DataUploadService")


@data_upload_service.main()
async def run(ctx: WorkflowContext) -> str:
    url = await ctx.run("bucket creation", create_s3_bucket)

    async def upload():
        await upload_data(url)

    await ctx.run("upload", upload)

    await ctx.promise("url").resolve(url)
    return url


@data_upload_service.handler("resultAsEmail")
async def result_as_email(ctx: WorkflowSharedContext, email: str):
    logging.info("Slow upload: client requested to be notified via email")
    url = await ctx.promise("url").value()

    async def send():
        await send_email(email, url)
    await ctx.run("email", send)


app = restate.app([data_upload_service])

File: java/patterns-use-cases/src/main/java/my/example/syncasync/DataUploadService.java
package my.example.syncasync;

import static my.example.syncasync.utils.DataOperations.createS3Bucket;
import static my.example.syncasync.utils.DataOperations.uploadData;

import dev.restate.sdk.Restate;
import dev.restate.sdk.annotation.Shared;
import dev.restate.sdk.annotation.Workflow;
import dev.restate.sdk.common.DurablePromiseKey;
import dev.restate.sdk.endpoint.Endpoint;
import dev.restate.sdk.http.vertx.RestateHttpServer;
import my.example.syncasync.utils.EmailClient;

@Workflow
public class DataUploadService {

  private static final DurablePromiseKey<String> URL_PROMISE =
      DurablePromiseKey.of("url", String.class);

  @Workflow
  public String run() {
    String url = Restate.run("create bucket", String.class, () -> createS3Bucket());
    Restate.run("upload", () -> uploadData(url));

    Restate.promiseHandle(URL_PROMISE).resolve(url);

    return url;
  }

  @Shared
  public void resultAsEmail(String email) {
    String url = Restate.promise(URL_PROMISE).future().await();
    Restate.run("send email", () -> EmailClient.send(url, email));
  }

  public static void main(String[] args) {
    RestateHttpServer.listen(Endpoint.bind(new DataUploadService()));
  }
}
